mongo/description/server_selector.go

```diff
@@ -182,31 +216,69 @@ func selectForReplicaSet(rp *readpref.ReadPref, t Topology, candidates []Server)
 		selected := selectByKind(candidates, RSPrimary)
 
 		if len(selected) == 0 {
-			selected = selectSecondaries(rp, candidates)
+			selected = selectSecondaries(rp, false, candidates)
 			return selectByTagSet(selected, rp.TagSets()), nil
 		}
 
 		return selected, nil
 	case readpref.SecondaryPreferredMode:
-		selected := selectSecondaries(rp, candidates)
+		selected := selectSecondaries(rp, false, candidates)
 		selected = selectByTagSet(selected, rp.TagSets())
 		if len(selected) > 0 {
 			return selected, nil
 		}
 		return selectByKind(candidates, RSPrimary), nil
 	case readpref.SecondaryMode:
-		selected := selectSecondaries(rp, candidates)
+		selected := selectSecondaries(rp, false, candidates)
 		return selectByTagSet(selected, rp.TagSets()), nil
 	case readpref.NearestMode:
 		selected := selectByKind(candidates, RSPrimary)
-		selected = append(selected, selectSecondaries(rp, candidates)...)
+		selected = append(selected, selectSecondaries(rp, false, candidates)...)
 		return selectByTagSet(selected, rp.TagSets()), nil
 	}
 
 	return nil, fmt.Errorf("unsupported mode: %d", rp.Mode())
 }
 
-func selectSecondaries(rp *readpref.ReadPref, candidates []Server) []Server {
+func selectForRSWithOutput(rp *readpref.ReadPref, t Topology, candidates []Server) ([]Server, error) {
+	if err := verifyMaxStaleness(rp, t); err != nil {
+		return nil, err
+	}
+
+	switch rp.Mode() {
+	case readpref.PrimaryMode:
+		return selectByKind(candidates, RSPrimary), nil
+	case readpref.PrimaryPreferredMode:
+		selected := selectByKind(candidates, RSPrimary)
+
+		if len(selected) == 0 {
+			selected = selectSecondaries(rp, true, candidates)
+			return selectByTagSet(selected, rp.TagSets()), nil
+		}
+
+		return selected, nil
+	case readpref.SecondaryMode, readpref.SecondaryPreferredMode:
```

Comment:
`selectForRSWithOutput` is very similar to `selectForReplicaSet`, except the `SecondaryMode` and `SecondaryPreferredMode` cases are collapsed. When executing an aggregate with an output stage, drivers must fall back to the primary when no usable secondaries are returned (i.e. `len(selected) == 0` on L268).

Reply:
Because `selectForRSWithOutput` is almost an exact copy of `selectForReplicaSet` except for the different behavior when selecting for an aggregate with output stage, I think it's better to modify `selectForReplicaSet` instead of adding a new function. 

E.g. possible modification:
```go
func selectForRSWithOutput(rp *readpref.ReadPref, t Topology, isOutputStage bool, candidates []Server) ([]Server, error) {
	// ...
	switch rp.Mode() {
	// ...
	case readpref.SecondaryMode:
		selected := selectSecondaries(rp, isOutputStage, candidates)
		selected = selectByTagSet(selected, rp.TagSets())
		// TODO: Comment describing why we're selecting primary
		// if there is an output stage.
		if len(selected) > 0 || !isOutputStage {
			return selected, nil
		}
		return selectByKind(candidates, RSPrimary), nil
	// ...
	}
}
```

Reply:
Great idea; done. Avoids some code duplication.